File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/grok/GrokLiveService.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.grok

import android.graphics.Bitmap
import android.util.Base64
import android.util.Log
import com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw.GrokToolCall
import com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw.GrokToolCallCancellation
import com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw.ToolDeclarations
import com.meta.wearable.dat.externalsampleapps.cameraaccess.settings.SettingsManager
import java.io.ByteArrayOutputStream
import java.time.Instant
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import org.json.JSONArray
import org.json.JSONObject

sealed class GrokConnectionState {
    data object Disconnected : GrokConnectionState()
    data object Connecting : GrokConnectionState()
    data object SettingUp : GrokConnectionState()
    data object Ready : GrokConnectionState()
    data class Error(val message: String) : GrokConnectionState()
}

class GrokLiveService {
    companion object {
        private const val TAG = "GrokLiveService"
    }

    private val _connectionState = MutableStateFlow<GrokConnectionState>(GrokConnectionState.Disconnected)
    val connectionState: StateFlow<GrokConnectionState> = _connectionState.asStateFlow()

    private val _isModelSpeaking = MutableStateFlow(false)
    val isModelSpeaking: StateFlow<Boolean> = _isModelSpeaking.asStateFlow()

    var onAudioReceived: ((ByteArray) -> Unit)? = null
    var onTurnComplete: (() -> Unit)? = null
    var onInterrupted: (() -> Unit)? = null
    var onDisconnected: ((String?) -> Unit)? = null
    var onInputTranscription: ((String) -> Unit)? = null
    var onOutputTranscription: ((String) -> Unit)? = null
    var onToolCall: ((GrokToolCall) -> Unit)? = null
    var onToolCallCancellation: ((GrokToolCallCancellation) -> Unit)? = null
    var onVisionSummary: ((String) -> Unit)? = null

    private var lastUserSpeechEnd: Long = 0
    private var responseLatencyLogged = false
    @Volatile private var visionRequestInFlight = false
    private val pendingToolCallIds = mutableSetOf<String>()
    @Volatile private var isAwaitingToolContinuation = false
    @Volatile private var hasToolResponseDone = false
    @Volatile private var cachedBrokerAccessToken: String? = null
    @Volatile private var cachedBrokerExpiresAtMs: Long = 0

    private var webSocket: WebSocket? = null
    private val sendExecutor = Executors.newSingleThreadExecutor()
    private var connectCallback: ((Boolean) -> Unit)? = null
    private var timeoutTimer: Timer? = null

    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .pingInterval(10, TimeUnit.SECONDS)
        .build()

    private val authClient = OkHttpClient.Builder()
        .readTimeout(15, TimeUnit.SECONDS)
        .connectTimeout(10, TimeUnit.SECONDS)
        .build()

    fun connect(callback: (Boolean) -> Unit) {
        val url = GrokConfig.websocketURL()
        if (url == null) {
            _connectionState.value = GrokConnectionState.Error("Grok auth not configured")
            callback(false)
            return
        }

        _connectionState.value = GrokConnectionState.Connecting
        resetToolContinuationState()
        connectCallback = callback

        sendExecutor.execute {
            val bearerToken = try {
                authorizationToken()
            } catch (e: Exception) {
                val msg = e.message ?: "Grok authorization failed"
                Log.e(TAG, msg)
                _connectionState.value = GrokConnectionState.Error(msg)
                resolveConnect(false)
                return@execute
            }

            val request = Request.Builder()
                .url(url)
                .addHeader("Authorization", "Bearer $bearerToken")
                .build()

            webSocket = client.newWebSocket(request, object : WebSocketListener() {
                override fun onOpen(webSocket: WebSocket, response: Response) {
                    Log.d(TAG, "WebSocket opened")
                    _connectionState.value = GrokConnectionState.SettingUp
                    sendSetupMessage()
                }

                override fun onMessage(webSocket: WebSocket, text: String) {
                    handleMessage(text)
                }

                override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                    handleMessage(bytes.utf8())
                }

                override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                    val msg = t.message ?: "Unknown error"
                    Log.e(TAG, "WebSocket failure: $msg")
                    _connectionState.value = GrokConnectionState.Error(msg)
                    _isModelSpeaking.value = false
                    resolveConnect(false)
                    onDisconnected?.invoke(msg)
                }

                override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                    Log.d(TAG, "WebSocket closing: $code $reason")
                    _connectionState.value = GrokConnectionState.Disconnected
                    _isModelSpeaking.value = false
                    resolveConnect(false)
                    onDisconnected?.invoke("Connection closed (code $code: $reason)")
                }

                override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                    Log.d(TAG, "WebSocket closed: $code $reason")
                    _connectionState.value = GrokConnectionState.Disconnected
                    _isModelSpeaking.value = false
                }
            })

            timeoutTimer = Timer().apply {
                schedule(object : TimerTask() {
                    override fun run() {
                        if (_connectionState.value == GrokConnectionState.Connecting
                            || _connectionState.value == GrokConnectionState.SettingUp) {
                            Log.e(TAG, "Connection timed out")
                            _connectionState.value = GrokConnectionState.Error("Connection timed out")
                            resolveConnect(false)
                        }
                    }
                }, 15000)
            }
        }
    }

    fun disconnect() {
        timeoutTimer?.cancel()
        timeoutTimer = null
        webSocket?.close(1000, null)
        webSocket = null
        onToolCall = null
        onToolCallCancellation = null
        resetToolContinuationState()
        _connectionState.value = GrokConnectionState.Disconnected
        _isModelSpeaking.value = false
        resolveConnect(false)
    }

    fun sendAudio(data: ByteArray) {
        if (_connectionState.value != GrokConnectionState.Ready) return
        sendExecutor.execute {
            webSocket?.send(JSONObject().apply {
                put("type", "input_audio_buffer.append")
                put("audio", Base64.encodeToString(data, Base64.NO_WRAP))
            }.toString())
        }
    }

    fun sendVideoFrame(bitmap: Bitmap) {
        if (_connectionState.value != GrokConnectionState.Ready) return
        if (!SettingsManager.visionSummariesEnabled) return
        sendExecutor.execute {
            val baos = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.JPEG, GrokConfig.VIDEO_JPEG_QUALITY, baos)
            summarizeFrame(baos.toByteArray())
        }
    }

    fun sendToolResponse(response: JSONObject) {
        val callId = response.optJSONObject("item")?.optString("call_id", "")
        sendExecutor.execute {
            webSocket?.send(response.toString())
        }
        if (!callId.isNullOrEmpty()) {
            synchronized(pendingToolCallIds) {
                pendingToolCallIds.remove(callId)
            }
        }
        sendResponseCreateIfToolCallsComplete()
    }

    fun sendTextMessage(text: String) {
        if (_connectionState.value != GrokConnectionState.Ready) return
        sendExecutor.execute {
            webSocket?.send(JSONObject().apply {
                put("type", "conversation.item.create")
                put("item", JSONObject().apply {
                    put("type", "message")
                    put("role", "user")
                    put("content", JSONArray().put(JSONObject().apply {
                        put("type", "input_text")
                        put("text", text)
                    }))
                })
            }.toString())
            webSocket?.send(JSONObject().put("type", "response.create").toString())
        }
    }

    private fun authorizationToken(): String {
        if (GrokConfig.hasAuthBroker) {
            return brokerAuthorizationToken()
        }
        if (GrokConfig.hasDirectAPIKey) {
            return GrokConfig.apiKey.trim()
        }
        throw IllegalStateException("Configure a Grok API key or Grok auth broker in Settings.")
    }

    private fun brokerAuthorizationToken(): String {
        val now = System.currentTimeMillis()
        val cached = cachedBrokerAccessToken
        if (!cached.isNullOrEmpty() && cachedBrokerExpiresAtMs - now > 60_000) {
            return cached
        }

        val brokerURL = GrokConfig.authBrokerURL.trim()
        if (brokerURL.isEmpty()) {
            throw IllegalStateException("Grok auth broker URL is empty.")
        }

        val requestBuilder = Request.Builder().url(brokerURL).get()
        val brokerToken = GrokConfig.authBrokerToken.trim()
        if (brokerToken.isNotEmpty() && brokerToken != "YOUR_OPENCLAW_GATEWAY_TOKEN") {
            requestBuilder.addHeader("Authorization", "Bearer $brokerToken")
        }

        authClient.newCall(requestBuilder.build()).execute().use { response ->
            val body = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw IllegalStateException(
                    "Grok auth broker returned HTTP ${response.code}: ${body.take(200)}"
                )
            }

            val json = JSONObject(body)
            val accessToken = json.optString("accessToken")
                .ifEmpty { json.optString("access_token") }
                .ifEmpty { json.optString("token") }
            if (accessToken.isEmpty()) {
                throw IllegalStateException("Grok auth broker did not return an access token.")
            }

            cachedBrokerAccessToken = accessToken
            cachedBrokerExpiresAtMs = parseExpiresAt(json)
            return accessToken
        }
    }

    private fun parseExpiresAt(json: JSONObject): Long {
        val expiresAt = json.opt("expiresAt") ?: json.opt("expires_at")
        if (expiresAt is Number) {
            return timestampToMillis(expiresAt.toLong())
        }
        if (expiresAt is String && expiresAt.isNotBlank()) {
            expiresAt.toLongOrNull()?.let { return timestampToMillis(it) }
            runCatching { Instant.parse(expiresAt).toEpochMilli() }.getOrNull()?.let { return it }
        }

        val expiresIn = json.opt("expiresIn") ?: json.opt("expires_in")
        val expiresInSeconds = when (expiresIn) {
            is Number -> expiresIn.toLong()
            is String -> expiresIn.toLongOrNull()
            else -> null
        } ?: 300
        return System.currentTimeMillis() + expiresInSeconds * 1000
    }

    private fun timestampToMillis(timestamp: Long): Long {
        return if (timestamp < 4_000_000_000L) timestamp * 1000 else timestamp
    }

    private fun resolveConnect(success: Boolean) {
        val cb = connectCallback
        connectCallback = null
        timeoutTimer?.cancel()
        timeoutTimer = null
        cb?.invoke(success)
    }

    private fun sendSetupMessage() {
        val setup = JSONObject().apply {
            put("type", "session.update")
            put("session", JSONObject().apply {
                put("model", GrokConfig.MODEL)
                put("instructions", GrokConfig.systemInstruction)
                put("voice", GrokConfig.voice)
                put("turn_detection", JSONObject().apply {
                    put("type", "server_vad")
                    put("threshold", 0.7)
                    put("prefix_padding_ms", 250)
                    put("silence_duration_ms", 650)
                })
                put("audio", JSONObject().apply {
                    put("input", JSONObject().apply {
                        put("format", JSONObject().apply {
                            put("type", "audio/pcm")
                            put("rate", GrokConfig.INPUT_AUDIO_SAMPLE_RATE)
                        })
                    })
                    put("output", JSONObject().apply {
                        put("format", JSONObject().apply {
                            put("type", "audio/pcm")
                            put("rate", GrokConfig.OUTPUT_AUDIO_SAMPLE_RATE)
                        })
                    })
                })
                put("tools", ToolDeclarations.allDeclarationsJSON())
            })
        }
        webSocket?.send(setup.toString())
    }

    private fun handleMessage(text: String) {
        try {
            val json = JSONObject(text)
            when (json.optString("type")) {
                "session.created",
                "conversation.created" -> Unit

                "session.updated" -> {
                    _connectionState.value = GrokConnectionState.Ready
                    resolveConnect(true)
                }

                "input_audio_buffer.speech_started" -> {
                    _isModelSpeaking.value = false
                    onInterrupted?.invoke()
                }

                "input_audio_buffer.speech_stopped" -> {
                    lastUserSpeechEnd = System.currentTimeMillis()
                    responseLatencyLogged = false
                }

                "conversation.item.input_audio_transcription.completed" -> {
                    val transcript = json.optString("transcript", "")
                    if (transcript.isNotEmpty()) {
                        Log.d(TAG, "You: $transcript")
                        onInputTranscription?.invoke(transcript)
                    }
                }

                "response.output_audio.delta" -> {
                    val base64Data = json.optString("delta", "")
                    if (base64Data.isNotEmpty()) {
                        markModelSpeakingIfNeeded()
                        onAudioReceived?.invoke(Base64.decode(base64Data, Base64.DEFAULT))
                    }
                }

                "response.output_audio_transcript.delta",
                "response.text.delta" -> {
                    val delta = json.optString("delta", "")
                    if (delta.isNotEmpty()) {
                        Log.d(TAG, "AI: $delta")
                        onOutputTranscription?.invoke(delta)
                    }
                }

                "response.function_call_arguments.done" -> {
                    GrokToolCall.fromJSON(json)?.let { toolCall ->
                        if (!isAwaitingToolContinuation) {
                            isAwaitingToolContinuation = true
                            hasToolResponseDone = false
                        }
                        synchronized(pendingToolCallIds) {
                            pendingToolCallIds.addAll(toolCall.functionCalls.map { it.id })
                        }
                        Log.d(TAG, "Tool call received: ${toolCall.functionCalls.size} function(s)")
                        onToolCall?.invoke(toolCall)
                    }
                }

                "response.done" -> {
                    _isModelSpeaking.value = false
                    responseLatencyLogged = false
                    if (isAwaitingToolContinuation) {
                        hasToolResponseDone = true
                        sendResponseCreateIfToolCallsComplete()
                    } else {
                        onTurnComplete?.invoke()
                    }
                }

                "error" -> {
                    val message = json.optJSONObject("error")?.optString("message")
                        ?: "Unknown Grok realtime error"
                    _connectionState.value = GrokConnectionState.Error(message)
                    _isModelSpeaking.value = false
                    resolveConnect(false)
                    onDisconnected?.invoke(message)
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error parsing message: ${e.message}")
        }
    }

    private fun markModelSpeakingIfNeeded() {
        if (_isModelSpeaking.value) return
        _isModelSpeaking.value = true
        if (lastUserSpeechEnd > 0 && !responseLatencyLogged) {
            val latency = System.currentTimeMillis() - lastUserSpeechEnd
            Log.d(TAG, "[Latency] ${latency}ms (user speech end -> first audio)")
            responseLatencyLogged = true
        }
    }

    private fun sendResponseCreateIfToolCallsComplete() {
        val isComplete = synchronized(pendingToolCallIds) { pendingToolCallIds.isEmpty() }
        if (!isAwaitingToolContinuation || !hasToolResponseDone || !isComplete) return

        isAwaitingToolContinuation = false
        hasToolResponseDone = false
        sendExecutor.execute {
            webSocket?.send(JSONObject().put("type", "response.create").toString())
        }
    }

    private fun resetToolContinuationState() {
        synchronized(pendingToolCallIds) {
            pendingToolCallIds.clear()
        }
        isAwaitingToolContinuation = false
        hasToolResponseDone = false
    }

    private fun summarizeFrame(jpegData: ByteArray) {
        if (visionRequestInFlight) return
        visionRequestInFlight = true
        try {
            val imageUrl = "data:image/jpeg;base64,${Base64.encodeToString(jpegData, Base64.NO_WRAP)}"
            val body = JSONObject().apply {
                put("model", GrokConfig.VISION_MODEL)
                put("store", false)
                put("temperature", 0.1)
                put("max_tokens", 120)
                put("messages", JSONArray()
                    .put(JSONObject().apply {
                        put("role", "system")
                        put("content", "You summarize live wearable-camera frames for a voice assistant. Return one compact, actionable sentence. Mention readable signs, people, hazards, screens, objects, or navigation cues. Do not invent certainty.")
                    })
                    .put(JSONObject().apply {
                        put("role", "user")
                        put("content", JSONArray()
                            .put(JSONObject().apply {
                                put("type", "image_url")
                                put("image_url", JSONObject().apply {
                                    put("url", imageUrl)
                                    put("detail", "low")
                                })
                            })
                            .put(JSONObject().apply {
                                put("type", "text")
                                put("text", "Summarize the current camera frame for the assistant.")
                            }))
                    }))
            }

            val request = Request.Builder()
                .url(GrokConfig.CHAT_COMPLETIONS_URL)
                .addHeader("Authorization", "Bearer ${authorizationToken()}")
                .addHeader("Content-Type", "application/json")
                .post(body.toString().toRequestBody("application/json; charset=utf-8".toMediaType()))
                .build()

            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return
                val responseBody = response.body?.string() ?: return
                val summary = JSONObject(responseBody)
                    .optJSONArray("choices")
                    ?.optJSONObject(0)
                    ?.optJSONObject("message")
                    ?.optString("content", "")
                    ?.trim()
                    .orEmpty()
                if (summary.isNotEmpty()) {
                    onVisionSummary?.invoke(summary)
                    injectVisionContext(summary)
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Vision summary failed: ${e.message}")
        } finally {
            visionRequestInFlight = false
        }
    }

    private fun injectVisionContext(summary: String) {
        if (_connectionState.value != GrokConnectionState.Ready) return
        webSocket?.send(JSONObject().apply {
            put("type", "conversation.item.create")
            put("item", JSONObject().apply {
                put("type", "message")
                put("role", "user")
                put("content", JSONArray().put(JSONObject().apply {
                    put("type", "input_text")
                    put("text", "Current visual context from the glasses camera: $summary")
                }))
            })
        }.toString())
    }
}

File: samples/CameraAccess/CameraAccess/Grok/GrokLiveService.swift
import Foundation
import UIKit

enum GrokConnectionState: Equatable {
  case disconnected
  case connecting
  case settingUp
  case ready
  case error(String)
}

@MainActor
class GrokLiveService: ObservableObject {
  @Published var connectionState: GrokConnectionState = .disconnected
  @Published var isModelSpeaking: Bool = false

  var onAudioReceived: ((Data) -> Void)?
  var onTurnComplete: (() -> Void)?
  var onInterrupted: (() -> Void)?
  var onDisconnected: ((String?) -> Void)?
  var onInputTranscription: ((String) -> Void)?
  var onOutputTranscription: ((String) -> Void)?
  var onToolCall: ((GrokToolCall) -> Void)?
  var onToolCallCancellation: ((GrokToolCallCancellation) -> Void)?
  var onVisionSummary: ((String) -> Void)?

  private var lastUserSpeechEnd: Date?
  private var responseLatencyLogged = false
  private var isVisionRequestInFlight = false
  private var pendingToolCallIds = Set<String>()
  private var isAwaitingToolContinuation = false
  private var hasToolResponseDone = false

  private var webSocketTask: URLSessionWebSocketTask?
  private var receiveTask: Task<Void, Never>?
  private var connectContinuation: CheckedContinuation<Bool, Never>?
  private let delegate = WebSocketDelegate()
  private let authProvider = GrokAuthProvider()
  private var urlSession: URLSession!
  private let sendQueue = DispatchQueue(label: "grok.send", qos: .userInitiated)

  init() {
    let config = URLSessionConfiguration.default
    config.timeoutIntervalForRequest = 30
    self.urlSession = URLSession(configuration: config, delegate: delegate, delegateQueue: nil)
  }

  func connect() async -> Bool {
    guard let url = GrokConfig.websocketURL() else {
      connectionState = .error("Invalid Grok WebSocket URL")
      return false
    }

    let bearerToken: String
    do {
      bearerToken = try await authProvider.authorizationToken()
    } catch {
      connectionState = .error(error.localizedDescription)
      return false
    }

    resetToolContinuationState()
    connectionState = .connecting

    let result = await withCheckedContinuation { (continuation: CheckedContinuation<Bool, Never>) in
      self.connectContinuation = continuation

      self.delegate.onOpen = { [weak self] _ in
        guard let self else { return }
        Task { @MainActor in
          self.connectionState = .settingUp
          self.sendSetupMessage()
          self.startReceiving()
        }
      }

      self.delegate.onClose = { [weak self] code, reason in
        guard let self else { return }
        let reasonStr = reason.flatMap { String(data: $0, encoding: .utf8) } ?? "no reason"
        Task { @MainActor in
          self.resolveConnect(success: false)
          self.connectionState = .disconnected
          self.isModelSpeaking = false
          self.onDisconnected?("Connection closed (code \(code.rawValue): \(reasonStr))")
        }
      }

      self.delegate.onError = { [weak self] error in
        guard let self else { return }
        let msg = error?.localizedDescription ?? "Unknown error"
        Task { @MainActor in
          self.resolveConnect(success: false)
          self.connectionState = .error(msg)
          self.isModelSpeaking = false
          self.onDisconnected?(msg)
        }
      }

      var request = URLRequest(url: url)
      request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")
      self.webSocketTask = self.urlSession.webSocketTask(with: request)
      self.webSocketTask?.resume()

      Task {
        try? await Task.sleep(nanoseconds: 15_000_000_000)
        await MainActor.run {
          self.resolveConnect(success: false)
          if self.connectionState == .connecting || self.connectionState == .settingUp {
            self.connectionState = .error("Connection timed out")
          }
        }
      }
    }

    return result
  }

  func disconnect() {
    receiveTask?.cancel()
    receiveTask = nil
    webSocketTask?.cancel(with: .normalClosure, reason: nil)
    webSocketTask = nil
    delegate.onOpen = nil
    delegate.onClose = nil
    delegate.onError = nil
    onToolCall = nil
    onToolCallCancellation = nil
    resetToolContinuationState()
    connectionState = .disconnected
    isModelSpeaking = false
    resolveConnect(success: false)
  }

  func sendAudio(data: Data) {
    guard connectionState == .ready else { return }
    enqueueJSON([
      "type": "input_audio_buffer.append",
      "audio": data.base64EncodedString()
    ])
  }

  func sendVideoFrame(image: UIImage) {
    guard connectionState == .ready else { return }
    guard SettingsManager.shared.visionSummariesEnabled else { return }
    guard let jpegData = image.jpegData(compressionQuality: GrokConfig.videoJPEGQuality) else { return }

    Task { [weak self] in
      await self?.summarizeFrame(jpegData: jpegData)
    }
  }

  func sendToolResponse(_ response: [String: Any]) {
    let callId = (response["item"] as? [String: Any])?["call_id"] as? String
    enqueueJSON(response)

    if let callId {
      pendingToolCallIds.remove(callId)
    }
    sendResponseCreateIfToolCallsComplete()
  }

  func sendTextMessage(_ text: String) {
    guard connectionState == .ready else { return }
    enqueueJSON([
      "type": "conversation.item.create",
      "item": [
        "type": "message",
        "role": "user",
        "content": [
          ["type": "input_text", "text": text]
        ]
      ]
    ])
    enqueueJSON(["type": "response.create"])
  }

  private func resolveConnect(success: Bool) {
    if let cont = connectContinuation {
      connectContinuation = nil
      cont.resume(returning: success)
    }
  }

  private func sendSetupMessage() {
    let setup: [String: Any] = [
      "type": "session.update",
      "session": [
        "model": GrokConfig.model,
        "instructions": GrokConfig.systemInstruction,
        "voice": GrokConfig.voice,
        "turn_detection": [
          "type": "server_vad",
          "threshold": 0.7,
          "prefix_padding_ms": 250,
          "silence_duration_ms": 650
        ],
        "audio": [
          "input": [
            "format": [
              "type": "audio/pcm",
              "rate": Int(GrokConfig.inputAudioSampleRate)
            ]
          ],
          "output": [
            "format": [
              "type": "audio/pcm",
              "rate": Int(GrokConfig.outputAudioSampleRate)
            ]
          ]
        ],
        "tools": ToolDeclarations.allDeclarations()
      ]
    ]
    enqueueJSON(setup)
  }

  private func enqueueJSON(_ json: [String: Any]) {
    guard let data = try? JSONSerialization.data(withJSONObject: json),
          let string = String(data: data, encoding: .utf8) else {
      return
    }
    let task = webSocketTask
    sendQueue.async {
      task?.send(.string(string)) { _ in }
    }
  }

  private func startReceiving() {
    receiveTask = Task { [weak self] in
      guard let self else { return }
      while !Task.isCancelled {
        guard let task = self.webSocketTask else { break }
        do {
          let message = try await task.receive()
          switch message {
          case .string(let text):
            await self.handleMessage(text)
          case .data(let data):
            if let text = String(data: data, encoding: .utf8) {
              await self.handleMessage(text)
            }
          @unknown default:
            break
          }
        } catch {
          if !Task.isCancelled {
            let reason = error.localizedDescription
            await MainActor.run {
              self.resolveConnect(success: false)
              self.connectionState = .disconnected
              self.isModelSpeaking = false
              self.onDisconnected?(reason)
            }
          }
          break
        }
      }
    }
  }

  private func handleMessage(_ text: String) async {
    guard let data = text.data(using: .utf8),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let type = json["type"] as? String else {
      return
    }

    switch type {
    case "session.created", "conversation.created":
      return

    case "session.updated":
      connectionState = .ready
      resolveConnect(success: true)

    case "input_audio_buffer.speech_started":
      isModelSpeaking = false
      onInterrupted?()

    case "input_audio_buffer.speech_stopped":
      lastUserSpeechEnd = Date()
      responseLatencyLogged = false

    case "conversation.item.input_audio_transcription.completed":
      if let transcript = json["transcript"] as? String, !transcript.isEmpty {
        NSLog("[Grok] You: %@", transcript)
        onInputTranscription?(transcript)
      }

    case "response.output_audio.delta":
      if let base64Data = json["delta"] as? String,
         let audioData = Data(base64Encoded: base64Data) {
        markModelSpeakingIfNeeded()
        onAudioReceived?(audioData)
      }

    case "response.output_audio_transcript.delta", "response.text.delta":
      if let delta = json["delta"] as? String, !delta.isEmpty {
        NSLog("[Grok] AI: %@", delta)
        onOutputTranscription?(delta)
      }

    case "response.output_audio_transcript.done", "response.text.done":
      return

    case "response.function_call_arguments.done":
      if let toolCall = GrokToolCall(json: json) {
        if !isAwaitingToolContinuation {
          isAwaitingToolContinuation = true
          hasToolResponseDone = false
        }
        for call in toolCall.functionCalls {
          pendingToolCallIds.insert(call.id)
        }
        NSLog("[Grok] Tool call received: %d function(s)", toolCall.functionCalls.count)
        onToolCall?(toolCall)
      }

    case "response.done":
      isModelSpeaking = false
      responseLatencyLogged = false
      if isAwaitingToolContinuation {
        hasToolResponseDone = true
        sendResponseCreateIfToolCallsComplete()
      } else {
        onTurnComplete?()
      }

    case "error":
      let error = json["error"] as? [String: Any]
      let message = error?["message"] as? String ?? "Unknown Grok realtime error"
      connectionState = .error(message)
      isModelSpeaking = false
      resolveConnect(success: false)
      onDisconnected?(message)

    default:
      return
    }
  }

  private func markModelSpeakingIfNeeded() {
    guard !isModelSpeaking else { return }
    isModelSpeaking = true
    if let speechEnd = lastUserSpeechEnd, !responseLatencyLogged {
      let latency = Date().timeIntervalSince(speechEnd)
      NSLog("[Latency] %.0fms (user speech end -> first audio)", latency * 1000)
      responseLatencyLogged = true
    }
  }

  private func sendResponseCreateIfToolCallsComplete() {
    guard isAwaitingToolContinuation,
          hasToolResponseDone,
          pendingToolCallIds.isEmpty else {
      return
    }

    isAwaitingToolContinuation = false
    hasToolResponseDone = false
    enqueueJSON(["type": "response.create"])
  }

  private func resetToolContinuationState() {
    pendingToolCallIds.removeAll()
    isAwaitingToolContinuation = false
    hasToolResponseDone = false
  }

  private func summarizeFrame(jpegData: Data) async {
    guard !isVisionRequestInFlight else { return }
    isVisionRequestInFlight = true
    defer { isVisionRequestInFlight = false }

    let bearerToken: String
    do {
      bearerToken = try await authProvider.authorizationToken()
    } catch {
      NSLog("[Grok] Vision auth failed: %@", error.localizedDescription)
      return
    }

    var request = URLRequest(url: GrokConfig.chatCompletionsURL)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")

    let imageURL = "data:image/jpeg;base64,\(jpegData.base64EncodedString())"
    let body: [String: Any] = [
      "model": GrokConfig.visionModel,
      "store": false,
      "temperature": 0.1,
      "max_tokens": 120,
      "messages": [
        [
          "role": "system",
          "content": "You summarize live wearable-camera frames for a voice assistant. Return one compact, actionable sentence. Mention readable signs, people, hazards, screens, objects, or navigation cues. Do not invent certainty."
        ],
        [
          "role": "user",
          "content": [
            [
              "type": "image_url",
              "image_url": [
                "url": imageURL,
                "detail": "low"
              ]
            ],
            [
              "type": "text",
              "text": "Summarize the current camera frame for the assistant."
            ]
          ]
        ]
      ]
    ]

    guard let httpBody = try? JSONSerialization.data(withJSONObject: body) else { return }
    request.httpBody = httpBody

    do {
      let (data, response) = try await URLSession.shared.data(for: request)
      guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
        return
      }
      guard
        let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
        let choices = object["choices"] as? [[String: Any]],
        let first = choices.first,
        let message = first["message"] as? [String: Any],
        let content = message["content"] as? String
      else {
        return
      }

      let summary = content.trimmingCharacters(in: .whitespacesAndNewlines)
      guard !summary.isEmpty else { return }
      onVisionSummary?(summary)
      injectVisionContext(summary)
    } catch {
      NSLog("[Grok] Vision summary failed: %@", error.localizedDescription)
    }
  }

  private func injectVisionContext(_ summary: String) {
    guard connectionState == .ready else { return }
    enqueueJSON([
      "type": "conversation.item.create",
      "item": [
        "type": "message",
        "role": "user",
        "content": [
          [
            "type": "input_text",
            "text": "Current visual context from the glasses camera: \(summary)"
          ]
        ]
      ]
    ])
  }
}

private class WebSocketDelegate: NSObject, URLSessionWebSocketDelegate {
  var onOpen: ((String?) -> Void)?
  var onClose: ((URLSessionWebSocketTask.CloseCode, Data?) -> Void)?
  var onError: ((Error?) -> Void)?

  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didOpenWithProtocol protocol: String?
  ) {
    onOpen?(`protocol`)
  }

  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
    reason: Data?
  ) {
    onClose?(closeCode, reason)
  }

  func urlSession(
    _ session: URLSession,
    task: URLSessionTask,
    didCompleteWithError error: Error?
  ) {
    if let error {
      onError?(error)
    }
  }
}
